zerver/migrations/0301_fix_unread_messages_in_deactivated_streams.py

```diff
@@ -1,34 +1,26 @@
-from django.db import connection, migrations
-from django.db.backends.postgresql.schema import DatabaseSchemaEditor
-from django.db.migrations.state import StateApps
+from django.db import migrations
 
 
-def mark_messages_read(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
-    Stream = apps.get_model("zerver", "Stream")
-    deactivated_stream_ids = list(Stream.objects.filter(deactivated=True).values_list('id', flat=True))
-    with connection.cursor() as cursor:
-        for i in deactivated_stream_ids:
-            cursor.execute(f"""
-                UPDATE zerver_usermessage SET flags = flags | 1
-                FROM zerver_message
-                INNER JOIN zerver_stream ON zerver_stream.recipient_id = zerver_message.recipient_id
-                WHERE zerver_message.id = zerver_usermessage.message_id
-                AND zerver_stream.id = {i};
-            """)
```

Comment:
The loop is likely important for servers that have large numbers of deactivated streams.  I suppose that's probably just zulip.com, which already had this migration, though?

Reply:
If I’m reading the `EXPLAIN` output correctly, PostgreSQL will do that loop on the outside anyway?

```console
$ ./manage.py populate_db -n 10000 --extra-streams 1000
$ psql zulip
zulip=> EXPLAIN UPDATE zerver_usermessage SET flags = flags | 1 FROM zerver_message INNER JOIN zerver_stream ON zerver_stream.recipient_id = zerver_message.recipient_id WHERE zerver_message.id = zerver_usermessage.message_id AND zerver_stream.deactivated;
                                                          QUERY PLAN                                                           
-------------------------------------------------------------------------------------------------------------------------------
 Update on zerver_usermessage  (cost=4.77..106.47 rows=146 width=46)
   ->  Nested Loop  (cost=4.77..106.47 rows=146 width=46)
         ->  Nested Loop  (cost=4.35..94.06 rows=8 width=16)
               ->  Seq Scan on zerver_stream  (cost=0.00..60.15 rows=1 width=10)
                     Filter: deactivated
               ->  Bitmap Heap Scan on zerver_message  (cost=4.35..33.83 rows=8 width=14)
                     Recheck Cond: (recipient_id = zerver_stream.recipient_id)
                     ->  Bitmap Index Scan on zerver_message_recipient_subject  (cost=0.00..4.34 rows=8 width=0)
                           Index Cond: (recipient_id = zerver_stream.recipient_id)
         ->  Index Scan using zerver_usermessage_message_id_f6e63a33 on zerver_usermessage  (cost=0.42..1.41 rows=10 width=30)
               Index Cond: (message_id = zerver_message.id)
(11 rows)
```

Reply:
Yeah, the issue is doing that whole operation in a single transaction, which is what happens without `atomic=False`; in our production it could result in holding locks for 20 minutes, which could effectively cause a user-facing outage.

In any case, this migration is already done in Zulip Cloud and I don't believe anyone else has enough data and downtime sensitivity that this would be a problem for them, so I think we can just merge.